src/inmanta/agent/handler.py

```diff
@@ -655,30 +594,103 @@ def do_changes(self, ctx: HandlerContext, resource: resources.Resource, changes:
+            return result.result["data"]
+
+        failed_dependencies = [req for req, status in requires.items() if status != ResourceState.deployed]
+        if not any(failed_dependencies):
+            self.execute(ctx, resource)
+            if _should_reload():
+                self.do_reload(ctx, resource)
+        else:
+            ctx.set_status(const.ResourceState.skipped)
+            ctx.info(
+                "Resource %(resource)s skipped due to failed dependency %(failed)s",
+                resource=resource.id.resource_version_str(),
+                failed=failed_dependencies,
+            )
+
+    def execute(self, ctx: HandlerContext, resource: resources.Resource) -> None:
         """
-        Update the given resource. This method is called by the agent. Most handlers will not override this method
+        Deploy the given resource. This method is called by the agent. Most handlers will not override this method
         and will only override :func:`~inmanta.agent.handler.ResourceHandler.check_resource`, optionally
         :func:`~inmanta.agent.handler.ResourceHandler.list_changes` and
         :func:`~inmanta.agent.handler.ResourceHandler.do_changes`
 
         :param ctx: Context object to report changes and logs to the agent and server.
         :param resource: The resource to check the current state of.
-        :param dry_run: True will only determine the required changes but will not execute them.
         """
         try:
             self.pre(ctx, resource)
 
             changes = self.list_changes(ctx, resource)
             ctx.update_changes(changes)
 
-            if not dry_run:
-                self.do_changes(ctx, resource, changes)
-                ctx.set_status(const.ResourceState.deployed)
+            self.do_changes(ctx, resource, changes)
+            ctx.set_status(const.ResourceState.deployed)
+        except SkipResource as e:
+            ctx.set_status(const.ResourceState.skipped)
+            ctx.warning(msg="Resource %(resource_id)s was skipped: %(reason)s", resource_id=resource.id, reason=e.args)
 
-            else:
-                ctx.set_status(const.ResourceState.dry)
+        except Exception as e:
+            ctx.set_status(const.ResourceState.failed)
+            ctx.exception(
+                "An error occurred during deployment of %(resource_id)s (exception: %(exception)s",
+                resource_id=resource.id,
+                exception=repr(e),
+            )
+        finally:
+            try:
+                self.post(ctx, resource)
+            except Exception as e:
+                ctx.exception(
+                    "An error occurred after deployment of %(resource_id)s (exception: %(exception)s",
+                    resource_id=resource.id,
+                    exception=repr(e),
+                )
+
+    def execute_dry_run(self, ctx: HandlerContext, resource: resources.Resource) -> None:
```

Comment:
I think this will need to call execute, but on the CRUD handler, it could be more specific. 
(I don't know if the crud handler calls list_changes normally?)


Perhaps this was a bad idea.

@bartv opinion? 
(my idea was to isolate the dryrun path from the normal execution path, to make it is very easy to ensure we don't accidentally execute, but perhaps that was a bad idea wrt to compatibility. )